### PR TITLE
[Tune] Fix checkpoint directory assignment for new checkpoints created after restoring a function trainable

### LINKED ISSUE #29947: [air] when restoring a session, new checkpoints saved seem to overwrite the old ones
### What happened + What you expected to happen

The following script should generate `ckpt_00000` in the first pass.
For 2nd pass, it should continue generating `ckpt_00001`. However, it overwrites `ckpt_00000`. In the end, there is only `ckpt_00000` in the folder when the trial terminates.

### Versions / Dependencies

master

### Reproduction script

```
import tensorflow as tf

import ray
from ray import tune
from ray.air import session
from ray.train.tensorflow.tensorflow_checkpoint import TensorflowCheckpoint


def to_tf_dataset(dataset, batch_size):
    def to_tensor_iterator():
        for batch in dataset.iter_tf_batches(
            batch_size=batch_size, dtypes=tf.float32
        ):
            yield tf.expand_dims(batch["x"], 1), batch["y"]

    output_signature = (
        tf.TensorSpec(shape=(None, 1), dtype=tf.float32),
        tf.TensorSpec(shape=(None), dtype=tf.float32),
    )
    tf_dataset = tf.data.Dataset.from_generator(
        to_tensor_iterator, output_signature=output_signature
    )
    return tf_dataset


def f(config):
    loaded_checkpoint = session.get_checkpoint()
    if loaded_checkpoint:
        is_restored = True
        # This should be loaded up from checkpoint. Should be doable after Balaji's change.
        start = 1  # continue training
        model = loaded_checkpoint.get_model()
    else:
        is_restored = False
        start = 0
        model = tf.keras.Sequential(
            [
                tf.keras.layers.InputLayer(input_shape=()),
                # Add feature dimension, expanding (batch_size,) to (batch_size, 1).
                tf.keras.layers.Flatten(),
                tf.keras.layers.Dense(1),
            ]
        )
    model.compile(optimizer="Adam", loss="mean_squared_error", metrics=["mse"])

    train_dataset = ray.data.from_items(
            [{"x": x, "y": x + 1} for x in range(32)])

    for epoch in range(start, 2):
        if not is_restored and epoch == 1:
            raise RuntimeError("Intentionally erroring out!!")
        tf_dataset = to_tf_dataset(dataset=train_dataset, batch_size=1)
        model.fit(tf_dataset)
        model.save("my_model")
        ckpt = TensorflowCheckpoint.from_saved_model("my_model")
        session.report({"epoch": epoch}, checkpoint=ckpt)

# tuner = tune.Tuner(f)  # use this as first pass
tuner = tune.Tuner.restore("/Users/xwjiang/ray_results/f_2022-11-02_09-24-21", resume_errored=True)  # use this as second pass
tuner.fit()

```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by justinvyu**
This PR fixes checkpoint directory creation for restored function trainables to use the restored iteration instead of starting over from `checkpoint_000000`.

## Why are these changes needed?

The `_StatusReporter` that handles checkpoint directory creation for function trainables keeps track of an `_iter` that closely follows the Trainable `training_iteration`, and it's used to create checkpoint directories (ex: `checkpoint_000000`).

Upon restoring a trial, this iteration is not restored properly and starts from 0, which results in a new checkpoint possibly overwriting an old one at the `checkpoint_000000` path.

The ticket below has more details, but the basic failure case:

- 1st checkpoint comes in, saved under `checkpoint_000000`
- Experiment is interrupted and gets restored
- 2nd checkpoint comes in, but still saved under `checkpoint_000000`
- 1st checkpoint is now overwritten

### Solution

Don't keep track of an `_iter` separately in the session - use the trainable's current `training_iteration` instead.

### Open Questions

1. Class trainables and function trainables have different indexing: Class trainables are 1-indexed with checkpoints matching the `training_iteration`, so the first checkpoint is saved as `checkpoint_000001`. Function trainables are 0-indexed. This PR could change it to make this consistent, and I think that the class trainable indexing makes more sense rather than always being 1-off with respect to the iteration number. **Decision: This would break backwards compatibility when restoring experiments that have 0 indexed checkpoints, so this will be left for a future PR.**

## Related issue number

Closes https://github.com/ray-project/ray/issues/29947

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/trainable/function_trainable.py [hunk 1]
```diff
@@ -123,33 +123,35 @@ def create_perm_checkpoint(checkpoint_dir, logdir, step):
...
-        trial_resources=None,
+        result_queue: queue.Queue,
+        continue_semaphore: threading.Semaphore,
+        end_event: threading.Event,
+        training_iteration_func: Callable[[], int],
+        experiment_name: Optional[str] = None,
+        trial_name: Optional[str] = None,
+        trial_id: Optional[str] = None,
+        logdir: Optional[str] = None,
+        trial_resources: Optional[Union[Resources, PlacementGroupFactory]] = None,
```
Comment (krfricke) [inline python/ray/tune/trainable/function_trainable.py:138]: Thanks!

File: python/ray/tune/trainable/function_trainable.py [hunk 2]
```diff
@@ -123,33 +123,35 @@ def create_perm_checkpoint(checkpoint_dir, logdir, step):
...
         self._last_checkpoint = None
         self._fresh_checkpoint = False
         self._trial_resources = trial_resources
-        # Also used as a marker of whether new `report()` API is being used,
-        # in which case, `_iter` will be incremented from 0 every time `report`
-        # is called.
-        self._iter = None
+        # Mark whether the `session.report()` API is being used,
+        # to throw an error if `tune.report()` is called as well
+        self._has_reported = False
```
Comment (krfricke) [inline python/ray/tune/trainable/function_trainable.py:154]: Is this "has reported using air.session.report"?

If so, can we slightly adjust the name to make this more obvious?

**Review by krfricke — APPROVED** [review]
Looks good, only one clarification

File: python/ray/tune/trainable/function_trainable.py (hunk 2 above)
Comment (justinvyu) [reply python/ray/tune/trainable/function_trainable.py:154]: Yes, this is for checking air session report usage. Will do!

**Pull request merged**